jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/DosHandler.java

```diff
@@ -0,0 +1,559 @@
+     */
+    public DosHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("maxRequestsPerSecond") int maxRequestsPerSecond,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, new ExponentialMovingAverageRateControlFactory(maxRequestsPerSecond), null, maxTrackers);
+    }
+
+    /**
+     * @param getId Function to extract an remote ID from a request.
+     * @param rateControlFactory Factory to create a Rate per Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DosHandler(
+        @Name("getId") Function<Request, String> getId,
+        @Name("rateFactory") RateControl.Factory rateControlFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        this(null, getId, rateControlFactory, rejectHandler, maxTrackers);
+    }
+
+    /**
+     * @param handler Then next {@link Handler} or {@code null}
+     * @param getId Function to extract an remote ID from a request.
+     * @param rateControlFactory Factory to create a Rate per Tracker
+     * @param rejectHandler A {@link Handler} used to reject excess requests, or {@code null} for a default.
+     * @param maxTrackers The maximum number of remote clients to track or -1 for a default value. If this limit is exceeded, then requests from additional remote clients are rejected.
+     */
+    public DosHandler(
+        @Name("handler") Handler handler,
+        @Name("getId") Function<Request, String> getId,
+        @Name("rateFactory") RateControl.Factory rateControlFactory,
+        @Name("rejectHandler") Request.Handler rejectHandler,
+        @Name("maxTrackers") int maxTrackers)
+    {
+        super(handler);
+        installBean(_trackers);
+        _getId = Objects.requireNonNullElse(getId, ID_FROM_REMOTE_ADDRESS);
+        installBean(_getId);
+        _rateControlFactory = Objects.requireNonNull(rateControlFactory);
+        installBean(_rateControlFactory);
+        _maxTrackers = maxTrackers <= 0 ? 10_000 : maxTrackers;
+        _rejectHandler = Objects.requireNonNullElseGet(rejectHandler, StatusRejectHandler::new);
+        installBean(_rejectHandler);
+    }
+
+    @Override
+    public void setServer(Server server)
+    {
+        super.setServer(server);
+        if (_rejectHandler instanceof Handler handler)
+            handler.setServer(server);
+    }
+
+    @Override
+    protected boolean onConditionsMet(Request request, Response response, Callback callback) throws Exception
+    {
+        // Reject if we have too many Trackers
+        if (_maxTrackers > 0 && _trackers.size() >= _maxTrackers)
```

Comment:
I think `_maxTrackers > 0` is always `true` because it is defaulted in the constructor.

Reply:
The intention was to have 0 being unlimited... but perhaps Integer.MAX_VALUE is better for that?   For now I've changed the constructor to allow a 0 value.

Reply:
The value `0` can also be used as an "emergency button", if it could be set via JMX, to stop all requests.
In this way, `0` is not special, -1 is default, and large values are large values.